binlog proto: add proto file

july2993 opened the pull request:
part of https://github.com/pingcap/tidb-tools/pull/21, contains only files about protobuf

siddontang commented:
LGTM

qiuyesuifeng commented:
LGTM

july2993 committed:
obinlog: add proto file

july2993 commented:
@GregoryIan PTAL

IANTHEREAL commented on obinlog/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+	optional bool is_null = 1 [default = false];
+
+    optional int64 int64_value = 2;
+	optional uint64 uint64_value = 3;
```
format code

IANTHEREAL commented on obinlog/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+}
+message DDLData {
+	// the current database use
+	optional string schema_name = 1;
```
how about adding table_name, if user want to route table.

IANTHEREAL replied:
add it

IANTHEREAL commented on obinlog/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+enum MutationType {
+	Insert = 0;
+	Update = 1;
+	Delete = 4;
```
2

IANTHEREAL commented on obinlog/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+    optional string  name       = 1 [(gogoproto.nullable) = false];
+    // column field type in mysql
+    optional string  mysql_type = 3 [(gogoproto.nullable) = false];
+	optional bool is_primary_key = 4 [(gogoproto.nullable) = false];
```
is_unique and is_primary?

IANTHEREAL replied:
the reason I say that is I think we try to restore it as much as possible

july2993 replied:
it's is_primary,  maybe we can add more info later

july2993 committed:
format file binlog.proto

july2993 committed:
generate new pb.go

iamxy commented:
Two questions first:
1. What does the `obinlog` mean?
2. What's the purpose of this PR?

Please follow the [Guidelines](https://github.com/pingcap/tidb/blob/master/pull_request_template.md) to describe the details about that.

july2993 commented:
> Two questions first:
>
> What does the obinlog mean?
> What's the purpose of this PR?
> Please follow the Guidelines to describe the details about that.

 just part of https://github.com/pingcap/tidb-tools/pull/21 for GregoryIan's request to easy review, add some describle in the relate pr, please see the relate pr

july2993 committed:
change file path

IANTHEREAL commented:
according to the discussion, the name of the binlog related module are:
* `binlog-service` =>  service that support binlog subscriptions
* `binlog-driver` => library that support binlog protocol parsing and reading (from kafka)
* `binlog-proto` => format protocol of binlog event

IANTHEREAL commented on obinlog/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+    optional int64 int64_value = 2;
+    optional uint64 uint64_value = 3;
+    optional double double_value = 4;
+    optional bytes bytes_value = 6;
```
5? check and fix similar problems

july2993 replied:
i think it's ok with this,  no need to use continuous number and should not change, but i can change it now as no servise use this proto now

IANTHEREAL replied:
I know it's ok,  but ugly

july2993 commented:
@WangXiangUSTC PATL and the relate PR

WangXiangUSTC commented on binlog_proto/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// for date and time type, string is set
```
how about change to "for date and time type,  set `string_value`"?

WangXiangUSTC commented on binlog_proto/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+// for blob and binary type, bytes is set
+// for enum, set, uint64 is set
+// for json, bytes is set
+message Column {
```
this struct name seems not suitable, how about change a name

july2993 replied:
field? i think column is ok

WangXiangUSTC commented on binlog_proto/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+message ColumnInfo {
+    optional string  name       = 1 [(gogoproto.nullable) = false];
+    // column field type in mysql
+    optional string  mysql_type = 2 [(gogoproto.nullable) = false];
```
how about add a enum to list the mysql type

july2993 replied:
i think we can just use the string same with mysql, but i will improve the comment

WangXiangUSTC commented on binlog_proto/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+message Binlog {
+    optional BinlogType    type           = 1 [(gogoproto.nullable) = false];
+    optional int64         commit_ts      = 2 [(gogoproto.nullable) = false];
+    // dml_data is marshalled from DML type,
```
',' -> '.'

WangXiangUSTC commented on binlog_proto/proto/binlog.proto:
```diff
@@ -0,0 +1,83 @@
+}
+
+enum BinlogType {
+    DML = 0; // has commit_ts, dml_data
```
what is the mean of "has commit_ts"

july2993 replied:
commit_ts field in the Binlog message, but seems no need to specify it, i will remove it

july2993 committed:
refine proto comment

july2993 commented:
@WangXiangUSTC  PTAL

WangXiangUSTC commented:
please format the code and comment, REST LGTM

july2993 committed:
format proto file use clang-format

july2993 commented:
@GregoryIan PTAL

july2993 committed:
proto/: introduce TableMutation

a txt has a mutation type sequence, so change it, or need to per table
per sequence item of tidb binlog

july2993 committed:
add table name for ddl

IANTHEREAL commented:
LGTM
